helpers/framed.js

```diff
@@ -0,0 +1,31 @@
+let framed;
+
+export async function isFramed() {
+	if (framed !== undefined) return framed;
+
+	if ((window.D2L && window.D2L.IsNotAnIframedApp) || window === window.parent) {
+		framed = false;
+		return framed;
+	}
+
+	const handleIsFramedResponse = (evt, resolve) => {
+		if (!evt || !evt.data || !evt.data.isFramed) return;
+		window.removeEventListener('message', evt => handleIsFramedResponse(evt, resolve), false);
+		framed = evt.data.isFramed;
+		resolve(framed);
+	};
+
+	framed = await Promise.race([
```

Comment:
The gist of this is that we resolve to the first promise that finishes. Thus, if we hit the 75ms timeout waiting for a response from the host page, we set our internal state false. However, if that listener is ever hit, we explicitly set `framed = evt.data.isFramed` so that the internal state will eventually update and future calls will start returning this new value.